crypto/x509/x509_vfy.c

```diff
@@ -973,28 +995,196 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
+            * that does not contain a response for one or more of the certificates
+            * in the completed certificate chain, the client SHOULD attempt to
+            * validate the certificate using an alternative retrieval method,
+            * such as downloading the relevant CRL;
+            */
+            if (crl_check_all_enabled || (crl_check_enabled && i == 0)) {
+                ok = check_cert_crl(ctx);
+                if (!ok)
+                    return ok;
+            } else {
+                ok = verify_cb_ocsp(ctx, X509_V_ERR_OCSP_VERIFY_FAILED);
+                if (!ok)
+                    return ok;
+            }
+        }
+
+        /*
+         * in the case that OCSP is only enabled for the server certificate
+         * and CRL for the complete chain, the rest of the chain has to be
+         * checked here
+         */
+        if (!ocsp_check_all_enabled && crl_check_all_enabled) {
+            last = sk_X509_num(ctx->chain) - 1;
+            for (i = 1; i <= last; i++) {
+                ctx->error_depth = i;
+                ok = check_cert_crl(ctx);
+                if (!ok)
+                    return ok;
+            }
+        }
+#endif
     } else {
-        /* If checking CRL paths this isn't the EE certificate */
-        if (ctx->parent != NULL)
-            return 1;
-        last = 0;
-    }
-    for (i = 0; i <= last; i++) {
-        ctx->error_depth = i;
-        ok = check_cert(ctx);
-        if (!ok)
-            return ok;
+        /*
+         * certificate status check with CRLs
+         */
+        if (crl_check_all_enabled) {
+            last = sk_X509_num(ctx->chain) - 1;
+        } else {
+            /* If checking CRL paths this isn't the EE certificate */
+            if (ctx->parent != NULL)
+                return 1;
+        }
+        for (i = 0; i <= last; i++) {
+            ctx->error_depth = i;
+            ok = check_cert_crl(ctx);
+            if (!ok)
+                return ok;
+        }
     }
     return 1;
 }
 
+#ifndef OPENSSL_NO_OCSP
+static int check_cert_ocsp_resp(X509_STORE_CTX *ctx)
+{
+    int cert_status, crl_reason;
+    int i, found = -1;
+    OCSP_BASICRESP *bs = NULL;
+    OCSP_RESPONSE *resp = NULL;
+    OCSP_CERTID *cert_id = NULL;
+    ASN1_GENERALIZEDTIME *rev, *thisupd, *nextupd;
+    int ret = V_OCSP_CERTSTATUS_UNKNOWN;
+
+    ctx->current_issuer = X509_find_by_subject(ctx->chain,
+                                               X509_get_issuer_name(ctx->current_cert));
+
+    if (sk_OCSP_RESPONSE_num(ctx->ocsp_resp) <= 0)
+        return X509_V_ERR_OCSP_NO_RESPONSE;
+
+    cert_id = OCSP_cert_to_id(NULL, ctx->current_cert, ctx->current_issuer);
+
+    for (i = 0; i < sk_OCSP_RESPONSE_num(ctx->ocsp_resp); i++) {
+        if ((resp = sk_OCSP_RESPONSE_value(ctx->ocsp_resp, i)) == NULL)
+            continue;
+
+        if ((bs = OCSP_response_get1_basic(resp)) == NULL)
+            continue;
+
+        found = OCSP_resp_find(bs, cert_id, -1);
```

Comment:
Should we make sure that this basic response actually was signed by the appropriate CA (issuer or designated OCSP/CRL signer?), otherwise OCSP responses from some unrelated (but also trusted) CA could be added to the OCSP response list, and trusted for status for certificates some other CA issued. Of course one should not trust such a rogue CA, but this also makes `cert_id` SHA1 collisions more significant, since now we need to care about cert_id collisions across CAs...